test/e2e_node/dockershim_checkpoint_test.go

```diff
@@ -0,0 +1,210 @@
+package e2e_node
+
+import (
+	"crypto/md5"
+	"fmt"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/test/e2e/framework"
+)
+
+const (
+	testPod                 = "checkpoint-test-pod"
+	testCheckpoint          = "checkpoint-test"
+	dockershimCheckpointDir = "/var/lib/dockershim/sandbox"
+	gcTimeout               = 3 * time.Minute
+	testCheckpointContent   = `{"version":"v1","name":"fluentd-gcp-v2.0-vmnqx","namespace":"kube-system","data":{},"checksum":1799154314}`
+)
+
+var _ = framework.KubeDescribe("Dockershim [Serial] [Disruptive]", func() {
+	f := framework.NewDefaultFramework("dockerhism-checkpoint-test")
+
+	It("should clean up pod sandbox checkpoint after pod deletion", func() {
+		runPodCheckpointTest(f, func() {
+			By("No-op manipulation")
+		})
+	})
+
+	It("should remove dangling checkpoint file", func() {
+		filename := fmt.Sprintf("%x", md5.Sum([]byte(fmt.Sprintf("%s/%s", testCheckpoint, f.Namespace.Name))))
+		fullpath := fmt.Sprintf("%s/%s", dockershimCheckpointDir, filename)
+
+		By(fmt.Sprintf("Write a file at %q", fullpath))
+		err := writeFileAndSync(fullpath, []byte(testCheckpointContent))
+		framework.ExpectNoError(err, "Failed to create file %q", fullpath)
+
+		By("Check if file is removed")
+		removed := false
+		for start := time.Now(); time.Since(start) < gcTimeout; time.Sleep(10 * time.Second) {
+			if _, err := os.Stat(fullpath); os.IsNotExist(err) {
+				removed = true
+				break
+			}
+		}
+		if !removed {
+			framework.Failf("Expect dangling checkpoint file to be removed")
+		}
+	})
+
+	Context("When pod sandbox checkpoint is missing", func() {
+		It("should complete pod sandbox clean up", func() {
+			runPodCheckpointTest(f, func() {
+				checkpoints := findCheckpoints(f.Namespace.Name)
+				if len(checkpoints) == 0 {
+					framework.Failf("No checkpoint for the pod was found")
+				}
+				By("Removing checkpoint of test pod")
+				for _, filename := range checkpoints {
+					filename = cleanUpOutput(filename)
+					if len(filename) == 0 {
+						continue
+					}
+					framework.Logf("Removing checkpiont %q", filename)
+					_, err := exec.Command("sudo", "rm", filename).CombinedOutput()
+					framework.ExpectNoError(err, "Failed to remove checkpoint file %q: %v", string(filename), err)
+				}
+			})
+		})
+	})
+
+	Context("When pod containers are missing", func() {
+		It("should complete pod sandbox clean up based on the information in sandbox checkpoint", func() {
+			runPodCheckpointTest(f, func() {
+				By("Gathering pod container ids")
+				stdout, err := exec.Command("sudo", "docker", "ps").CombinedOutput()
```

Comment:
There are better way to get cleaner output. `-f` lets you filter containers by key-values. `-q` returns only container IDs.
E.g., to return the container IDs of all containers whose name includes `kube-system`:
`sudo docker ps -f name=kube-system -q`